### DISCUSSION TITLE
Get only the dirty field values

### DISCUSSION THREAD

**Discussion by dandv** [react-hook-form/react-hook-form#1991] [Q&A] [answered]
Is there a direct way to get dirty field *values*? This is useful when submitting updates to an API endpoint - only transmit over the wire the changed field(s), not the entire object/form.

[`formState.dirtyFields`](https://react-hook-form.com/api#formState) is an object with `true` values for the fields, and that could be used to match over the [handleSubmit](https://react-hook-form.com/api#handleSubmit) `data`, but doing that properly takes recursion, and before I implement it by hand, I wanted to check if this is already provided more directly. A parameter to `getValues` would be great.

**Comment by bluebill1049** [comment]
hey @dandv we don't have such mechanism, it's easier to implement at your App level.

**Reply by platoonic** [reply]
Hey @bluebill1049
I think it would be pretty cool if RHF had an optional argument for "handleSubmit" that submits only touched (dirty) values,
It's a quite needed thing for sending requests to a PUT endpoints, What are your thoughts?

**Reply by bluebill1049** [reply]
looks like I have missed this message, feel free to send a feature request on this.

**Comment by dandv** [comment, marked as answer]
OK, this was pretty easy to implement in a generic way:

```typescript
// Map RHF's dirtyFields over the `data` received by `handleSubmit` and return the changed subset of that data.
export function dirtyValues(dirtyFields: object | boolean, allValues: object): object {
  // If *any* item in an array was modified, the entire array must be submitted, because there's no way to indicate
  // "placeholders" for unchanged elements. `dirtyFields` is `true` for leaves.
  if (dirtyFields === true || Array.isArray(dirtyFields))
    return allValues;
  // Here, we have an object
  return Object.fromEntries(Object.keys(dirtyFields).map(key => [key, dirtyValues(dirtyFields[key], allValues[key])]));
}
```

Code sandbox with tests at https://codesandbox.io/s/ts-jest-34w4x?file=/src/index.ts.

I think this would be a useful feature to add to `getValues` behind an option. I'll let you decide if it's worth it and adjust the types.

**Reply by bluebill1049** [reply]
Thanks @dandv for sharing the solution.

**Reply by GollyJer** [reply]
@dandv This is a great function. Thanks for sharing! 🎉🙏

To give back... slightly. Here's a version with updated `types` for more recent versions of typescript.

```javascript
type UnknownArrayOrObject = unknown[] | Record<string, unknown>;

export const dirtyValues = (
  dirtyFields: UnknownArrayOrObject | boolean,
  allValues: UnknownArrayOrObject
): UnknownArrayOrObject => {
  // NOTE: Recursive function.

  // If *any* item in an array was modified, the entire array must be submitted, because there's no
  // way to indicate "placeholders" for unchanged elements. `dirtyFields` is `true` for leaves.
  if (dirtyFields === true || Array.isArray(dirtyFields)) {
    return allValues;
  }

  // Here, we have an object.
  return Object.fromEntries(
    Object.keys(dirtyFields).map((key) => [key, dirtyValues(dirtyFields[key], allValues[key])])
  );
};
```

**Reply by revskill10** [reply]
@GollyJer  How to use this in `onSubmit` ?

**Reply by ghost** [reply]
@GollyJer TS still doesn't like this: 
```dirtyValues(dirtyFields[key], allValues[key])])```

returns:

```Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'false | Record<string, unknown>'.```
 ```No index signature with a parameter of type 'string' was found on type 'false | Record<string, unknown>'.```

**Reply by revich2** [reply]
@revskill10 you can do something like this:
```ts
const dirtyFields = useRef<DeepMap<TestFormValues, true>>();
useEffect(() => {
  dirtyFields.current = formState.dirtyFields;
});
```
and use `dirtyFields.current` in your `onSubmit` function

**Reply by nicholascm** [reply]
This is a really useful method - any reason it shouldn't be exported from `react-hook-form` at least as a utility that is available ? I could open a PR if desired - it looks like all the work has been done here by others already.

**Reply by bluebill1049** [reply]
thanks @nicholascm perhaps we can include an example in the examples folder?

**Reply by nicholascm** [reply]
https://github.com/react-hook-form/react-hook-form/pull/6662

**Reply by dandv** [reply]
I think this should be in core.

**Reply by bluebill1049** [reply]
wouldn't mind having a feature request and put that into a vote for the upvote. if the demand is high, we can look into an API with RFC.

**Reply by npetridis** [reply]
Any update on this one? is there a link where we can vote for the feature? thank you!

**Reply by Moshyfawn** [reply]
I doubt there was any movement on the matter, as the feature request hasn't been made to [the project's board](https://github.com/react-hook-form/react-hook-form/projects/1).
If you're interested in seeing this feature, you can create [a feature request](https://github.com/react-hook-form/react-hook-form/issues/new?assignees=&labels=&template=feature_request.md&title=)

**Reply by zaverden** [reply]
this approach breaks the best part of hook-form: whole form does not have a rerender on field change. but here we are, subscribing to `formState.dirtyFields` at the root of the form, and get that rerender back (on the fist field change).

```
const { getActualFormState, handleSubmit, .... } = useForm(...)

const onSubmit = handleSubmit((values) => saveDirty(values, getActualFormState().dirtyFields));
```

with extra `getActualFormState` method from `useForm` it whould work just fine

**Reply by zaverden** [reply]
[workaround](https://codesandbox.io/s/react-hook-form-submit-only-dirty-fields-without-rerender-6esie2?file=/src/index.tsx:162-754):

`ReversePartialFormStateProvider` is a child component. is subscribes to the required part of the `formState` via `useFormeState` using `control`. and sends updated state via ref to the parent.

well, it is quiet ugly, but it works without whole form rerender.

(I have really huge form, like 300+ fields. it is nice to render it just once)

**Reply by MoatezNG** [reply]
Small modification incase someone needs it with an array of objects : 

```
type UnknownArrayOrObject = unknown[] | Record<string, unknown>;

export const dirtyValues = (
  dirtyFields: UnknownArrayOrObject | boolean,
  allValues: UnknownArrayOrObject
): UnknownArrayOrObject => {
  if (Array.isArray(dirtyFields)) {
    const dirtyArray = dirtyFields.map((el, index) => {
      return Object.fromEntries(
        Object.keys(el).map((key) => [key, allValues[index][key]])
      );
    });
    return dirtyArray;
  }
  return Object.fromEntries(
    Object.keys(dirtyFields).map((key) => [
      key,
      dirtyValues(dirtyFields[key], allValues[key]),
    ])
  );
};

// how to use it with onSubmit 
onClick={handleSubmit((values) => {
                onSumbit(dirtyValues(dirtyFields, values));
              })}
```

**Reply by gremo** [reply]
@MoatezNG your code doesn't work right, you need to add che `true `check before `Array.isArray`.

**Reply by pedrollmatias** [reply]
> @GollyJer TS still doesn't like this: `dirtyValues(dirtyFields[key], allValues[key])])`
> 
> returns:
> 
> `Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'false | Record<string, unknown>'.` `No index signature with a parameter of type 'string' was found on type 'false | Record<string, unknown>'.`

@ghost A code that typescript likes:
```ts
type UnknownObject = Record<string, unknown>;
type UnknownArrayOrObject = unknown[] | UnknownObject;

const dirtyFields = (
  dirtyFields: UnknownArrayOrObject | boolean | unknown,
  allValues: UnknownArrayOrObject | unknown
): UnknownArrayOrObject | unknown => {
  if (dirtyFields === true || Array.isArray(dirtyFields)) {
    return allValues;
  }

  const dirtyFieldsObject = dirtyFields as UnknownObject;
  const allValuesObject = allValues as UnknownObject;

  return Object.fromEntries(
    Object.keys(dirtyFieldsObject).map((key) => [key, dirtyFields(dirtyFieldsObject[key], allValuesObject[key])])
  );
};
```

**Reply by dannobytes** [reply]
An alternative that uses a "reducer" to ensure the function is "pure" and has some inferred typings to ensure that `values` is an all-inclusive superset of `dirtyFields`, i.e. ensures that all dirty fields has a corresponding values match.

```tsx
function getDirtyValues<DirtyFields extends Record<string, unknown>, Values extends Record<keyof DirtyFields, unknown>>(
  dirtyFields: DirtyFields,
  values: Values
): Partial<typeof values> {
  const dirtyValues = Object.keys(dirtyFields).reduce((prev, key) => {
    // Unsure when RFH sets this to `false`, but omit the field if so.
    if (!dirtyFields[key]) return prev;

    return {
      ...prev,
      [key]:
        typeof dirtyFields[key] === 'object'
          ? getDirtyValues(dirtyFields[key] as DirtyFields, values[key] as Values)
          : values[key],
    };
  }, {});

  return dirtyValues;
}
```

**Reply by Szorcu** [reply]
@dannobytes I love your solution. Here is my version of your approach :slightly_smiling_face:  

```typescript
export const getItemsDirtyData = <
  TData extends Record<keyof TDirtyItems, unknown>,
  TDirtyItems extends Record<string, unknown>
>(
  data: TData,
  dirtyItems: TDirtyItems
): Partial<TData> => {
  const dirtyItemsEntries = Object.entries(dirtyItems)

  return dirtyItemsEntries.reduce((dirtyData, [name, value]) => {
    if (typeof value !== "object") {
      return { ...dirtyData, [name]: data[name] }
    }

    return {
      ...dirtyData,
      [name]: getItemsDirtyData(
        data[name] as TData,
        dirtyItems[name] as TDirtyItems
      ),
    }
  }, {})
}
```

**Reply by brian-gates** [reply]
Yet another variation.  This one returns only the keys of values which are dirty, with a value of true. Basically the same API as `dirtyFields` but with clean fields omitted.

```ts
export const getDirtyFields = <
  TData extends Record<keyof TDirtyItems, unknown>,
  TDirtyItems extends Record<string, unknown>,
>(
  formValues: TData,
  dirtyItems: TDirtyItems,
): Partial<TData> => {
  return Object.entries(dirtyItems).reduce((dirtyData, [key, value]) => {
    if (value === false) return dirtyData;
    if (value === true) return { ...dirtyData, [key]: true };

    const child = getDirtyFields(
      formValues[key] as TData,
      dirtyItems[key] as TDirtyItems,
    );

    if (typeof child === 'object' && Object.keys(child).length === 0) {
      return dirtyData;
    }

    if (Array.isArray(child) && child.length === 0) {
      return dirtyData;
    }

    return {
      ...dirtyData,
      [key]: child,
    };
  }, {});
};
```

**Reply by CamiloTello002** [reply]
@dannobytes Man, thank you so much!!! I copy pasted your solution and it worked flawlessly. :-)

**Reply by zcmgyu** [reply]
Code created by [dannobytes](https://github.com/dannobytes) did not pass test cases created by [dandv](https://github.com/dandv)

This is my version which passed all testcase

```typescript
export type DirtyFieldsType =
  | boolean
  | null
  | {
      [key: string]: DirtyFieldsType;
    }
  | DirtyFieldsType[];

export function getDirtyValues<T extends Record<string, any>>(
  dirtyFields: Partial<Record<keyof T, DirtyFieldsType>>,
  values: T,
): Partial<T> {
  const dirtyValues = Object.keys(dirtyFields).reduce((prev, key) => {
    const value = dirtyFields[key];
    if (!value) {
      return prev;
    }
    const isObject = typeof value === 'object';
    const isArray = Array.isArray(value);
    const nestedValue =
      isObject && !isArray
        ? getDirtyValues(value as Record<string, any>, values[key])
        : values[key];
    return { ...prev, [key]: isArray ? values[key] : nestedValue };
  }, {} as Partial<T>);
  return dirtyValues;
}
```

**Reply by ValerioDContinuum** [reply]
> Code created by [dannobytes](https://github.com/dannobytes) did not pass test cases created by [dandv](https://github.com/dandv)
> 
> This is my version which passed all testcase
> 
> ```ts
> export type DirtyFieldsType =
>   | boolean
>   | null
>   | {
>       [key: string]: DirtyFieldsType;
>     }
>   | DirtyFieldsType[];
> 
> export function getDirtyValues<T extends Record<string, any>>(
>   dirtyFields: Partial<Record<keyof T, DirtyFieldsType>>,
>   values: T,
> ): Partial<T> {
>   const dirtyValues = Object.keys(dirtyFields).reduce((prev, key) => {
>     const value = dirtyFields[key];
>     if (!value) {
>       return prev;
>     }
>     const isObject = typeof value === 'object';
>     const isArray = Array.isArray(value);
>     const nestedValue =
>       isObject && !isArray
>         ? getDirtyValues(value as Record<string, any>, values[key])
>         : values[key];
>     return { ...prev, [key]: isArray ? values[key] : nestedValue };
>   }, {} as Partial<T>);
>   return dirtyValues;
> }
> ```

To avoid having `any` in my project, is it a good solution to use type `FieldValues` from RHF instead of `Record<string, any>`?

TS suggest me to use `unknown` instead of `any` but I then get these TS errors, depending on where I use it or not:

![Screenshot 2025-05-05 alle 17 16 47](https://github.com/user-attachments/assets/7377a565-211e-47fd-b7cc-27b0994d44b9)

![Screenshot 2025-05-05 alle 17 17 02](https://github.com/user-attachments/assets/7d918d4a-89d9-4bfa-ace1-4bfe5967d01d)

**Comment by Ali-Hussein-dev** [comment]
the recommend new API is pretty useful in many cases. but I prefer another name to use that is a bit intuitive like `getModifiedFields` or `modifiedFields`

**Reply by Moshyfawn** [reply]
"Dirty" input is the common way to refer to an input which value has been modified and is different from it's default value.

**Comment by tykhan** [comment]
That an alternate version of helper, which gives you perfect TS autocomplete in your components

```
import { FieldValues } from 'react-hook-form';

export const filterChangedFormFields = <T extends FieldValues>(
    allFields: T,
    dirtyFields: Partial<Record<keyof T, boolean>>
): Partial<T> => {
    const changedFieldValues = Object.keys(dirtyFields).reduce((acc, currentField) => {
        return {
            ...acc,
            [currentField]: allFields[currentField]
        }
    }, {} as Partial<T>);

    return changedFieldValues;
};
```

**Reply by lveillard** [reply]
So in my case  i have an array of strings which generates a ts error. Adapting your code to this

```
export const filterChangedFormFields = <T extends FieldValues>(
  allFields: T,
  dirtyFields: Partial<Record<keyof T, boolean | boolean[]>>
): Partial<T> => {
  const changedFieldValues = Object.keys(dirtyFields).reduce(
    (acc, currentField) => {
      const isDirty = Array.isArray(dirtyFields[currentField])
        ? (dirtyFields[currentField] as boolean[]).some((value) => value === true)
        : dirtyFields[currentField] === true;
      if (isDirty) {
        return {
          ...acc,
          [currentField]: allFields[currentField],
        };
      }
      return acc;
    },
    {} as Partial<T>
  );

  return changedFieldValues;
};
```
Fixes it


Also i agree,,, that this should be a feature getDirtyFields() or something like this

**Reply by minmaxdata** [reply]
these last examples arent recursive tho

**Comment by ghost** [comment]
I think using dirty fields is not enough so [I'm proposing a convention](https://github.com/orgs/react-hook-form/discussions/11008) to extract only the data that has been modified from the form's state. What do you think about it?

**Comment by rudrava-myhomepathway** [comment]
```

const cleanData = data => Object.entries(data)
    .filter(([key, value]) => value !== undefined)
    .reduce((obj, [key, value]) => {
        obj[key] = value;
        return obj;
    }, {});


const dirtyValues = (dirtyFields, allValues) => {
    if (dirtyFields === true || Array.isArray(dirtyFields))
        return allValues;
    // Here, we have an object
    return cleanData(Object.fromEntries(Object.keys(dirtyFields).map(key => [key, dirtyFields[key] ? dirtyValues(dirtyFields[key], allValues[key]) : undefined])));
}


```

you can use this if you want just the field swith the data
